compiler_opt/tools/bc_hyperparam_tuning.py

```diff
@@ -0,0 +1,92 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script for testing different hyperparameter combinations for BC training."""
+
+import os
+
+from absl import app
+from absl import flags
+import tempfile
+import gin
+
+from compiler_opt.rl import train_bc
+from compiler_opt.tools import generate_default_trace
+
+import tensorflow as tf
+from tensorboard.plugins.hparams import api as hp
+
+_DATA_PATH = flags.DEFINE_multi_string(
+    'data_path', [],
+    'Path to TFRecord file(s) containing training data. Skip training and dump'
+    'an untrained model with random weights (for testing purpose) if '
+    'unspecified.')
+_CORPUS_PATH = flags.DEFINE_string(
+    'corpus_path', '', 'Path to the corpus to evaluate the models.')
+_OUTPUT_DIR = flags.DEFINE_string('output_dir', '', 'Path to the output data')
+_GIN_FILES = flags.DEFINE_multi_string(
+    'gin_files', [], 'List of paths to gin configuration files.')
+_GIN_BINDINGS = flags.DEFINE_multi_string(
+    'gin_bindings', [],
+    'Gin bindings to override the values set in the config files.')
+
+METRIC_REWARD = 'reward'
+
+
+def run_training():
+  with tempfile.TemporaryDirectory() as temp_dir:
+    root_dir = os.path.expanduser(temp_dir)
+    root_dir = os.path.normpath(root_dir)
+    # TODO(boomanaiden154): leave num_iterations and batch_size to config
+    train_bc.train_eval(
+        num_iterations=100,
+        batch_size=64,
+        root_dir=root_dir,
+        data_path=_DATA_PATH.value)
+    return generate_default_trace.generate_default_trace(
+        data_path=_CORPUS_PATH.value,
+        policy_path=os.path.join(root_dir, './saved_policy'),
+        sampling_rate=0.01)
+
+
+def run_tuning():
+  hp_num_units_layer1 = hp.HParam('num_units_layer1', hp.Discrete([40, 80]))
+  hp_num_units_layer2 = hp.HParam('num_units_layer2', hp.Discrete([40, 80]))
+  run_number = 0
+  for num_units_layer1 in hp_num_units_layer1.domain.values:
+    for num_units_layer2 in hp_num_units_layer2.domain.values:
+      hparams = {
+          'HP_NUM_UNITS_LAYER1': num_units_layer1,
+          'HP_NUM_UNITS_LAYER2': num_units_layer2
+      }
+      with gin.unlock_config():
+        gin.bind_parameter('RegAllocNetwork.fc_layer_params',
+                           (num_units_layer1, num_units_layer2))
+      output_path = os.path.join(_OUTPUT_DIR.value, f'./run{run_number}')
+      mean_reward = run_training()
```

Comment:
if i understand it correctly, it only runs the next set of hyper parameters after the current one is finished? so one after another?

Reply:
Yes. It runs them all sequentially. Running in them in parallel on the same machine probably wouldn't provide any speedup as the BC training is already heavily multithreaded. Running it in parallel distributed across multiple machines would help, but supporting that doesn't seem necessary currently.